conn.go

```diff
@@ -262,33 +357,125 @@ func connect(config ConnConfig, connInfo *pgtype.ConnInfo) (c *Conn, err error)
 		}
 	}
 
+	if c.config.TargetSessionAttrs != "" &&
+		c.config.TargetSessionAttrs != "any" &&
+		c.config.TargetSessionAttrs != "read-write" {
+		return nil, errors.New("invalid value for target_session_attrs, expected \"any\" or \"read-write\"")
+	}
+
 	c.onNotice = config.OnNotice
 
-	network, address := c.config.networkAddress()
 	if c.config.Dial == nil {
 		d := defaultDialer()
 		c.config.Dial = d.Dial
 	}
 
-	if c.shouldLog(LogLevelInfo) {
-		c.log(LogLevelInfo, "Dialing PostgreSQL server", map[string]interface{}{"network": network, "address": address})
+	addrs, err := c.config.networkAddresses()
+	if err != nil {
+		return nil, err
 	}
-	err = c.connect(config, network, address, config.TLSConfig)
-	if err != nil && config.UseFallbackTLS {
+
+	var errs []error
+	for _, addr := range addrs {
+		network, address := addr.Network(), addr.String()
+
 		if c.shouldLog(LogLevelInfo) {
-			c.log(LogLevelInfo, "connect with TLSConfig failed, trying FallbackTLSConfig", map[string]interface{}{"err": err})
+			c.log(LogLevelInfo, "Dialing PostgreSQL server", map[string]interface{}{
+				"network": network,
+				"address": address,
+			})
+		}
+
+		err = c.connect(config, network, address, config.TLSConfig)
+		if err != nil && config.UseFallbackTLS {
+			if c.shouldLog(LogLevelInfo) {
+				c.log(LogLevelInfo, "connect with TLSConfig failed, trying FallbackTLSConfig", map[string]interface{}{
+					"err":     err,
+					"network": network,
+					"address": address,
+				})
+			}
+			err = c.connect(config, network, address, config.FallbackTLSConfig)
+		}
+
+		if err != nil {
+			if c.shouldLog(LogLevelError) {
+				c.log(LogLevelError, "connect failed", map[string]interface{}{
+					"err":     err,
+					"network": network,
+					"address": address,
+				})
+			}
+
+			// On any auth errors return immediately
+			if pgErr, ok := err.(PgError); ok {
+				switch pgErr.Code {
+				// @see: https://www.postgresql.org/docs/current/errcodes-appendix.html
+				case "28000", "28P01": // Invalid Authorization Specification
+					return nil, pgErr
+				}
+			}
+
+			errs = append(errs, err)
+			continue
+		}
+
+		err = c.writable()
+		if err != nil {
+			c.die(err)
+
+			if c.shouldLog(LogLevelInfo) {
+				c.log(LogLevelInfo, "host is not writable", map[string]interface{}{
+					"err":     err,
+					"network": network,
+					"address": address,
+				})
+			}
+
+			errs = append(errs, err)
+			continue
 		}
-		err = c.connect(config, network, address, config.FallbackTLSConfig)
+
+		c.addr = addr
+
+		return c, nil
+	}
+
+	// To keep backwards compatibility, if specific error type expected.
+	if len(errs) == 1 {
+		return nil, errs[0]
+	}
+
+	var errmsg string
+	for _, err := range errs {
+		errmsg += "; " + err.Error()
+	}
+
+	return nil, errors.New(errmsg)
+}
+
+func (c *Conn) writable() error {
```

Comment:
m.b. rename to checkWritableIfReuired()?
or just checkSessionAtrs()?

Reply:
`checkWritableIfReuired` - looks little bit java-style :) `checkWritable` maybe?

Reply:
I believe we do not check writability here but check compliance with target session attributes. I think we could also restructure code a bit to make the future implementation of prefer-standby simpler.

Reply:
https://www.postgresql.org/docs/devel/libpq-connect.html#libpq-connstring

> When multiple hosts are specified, or when a single host name is translated to multiple addresses, all the hosts and addresses will be tried in order, until one succeeds. If none of the hosts can be reached, the connection fails. If a connection is established successfully, but authentication fails, the remaining hosts in the list are not tried.

Reply:
I got you point, but want to leave this function called `checkWritable`, cuz all other options don't helps to read and understand code.